tests/commands/test_store_commands.py

```diff
@@ -955,6 +966,25 @@ def test_upload_call_error(emitter, store_mock, config, tmp_path, formatted):
         emitter.assert_messages(expected)
 
 
+@pytest.mark.parametrize("formatted", [None, JSON_FORMAT])
+def test_upload_call_login_expired(monkeypatch, config, tmp_path, formatted):
+    """Simple upload but login expired."""
+    monkeypatch.setenv("CHARMCRAFT_AUTH", base64.b64encode("credentials".encode()).decode())
+    with patch("craft_store.base_client.HTTPClient.request") as mock_whoami:
+        mock_whoami.side_effect = StoreServerError(_fake_response(401, json={}))
+
+        test_charm = tmp_path / "mystuff.charm"
+        _build_zip_with_yaml(test_charm, "metadata.yaml", content={"name": "mycharm"})
+        args = Namespace(filepath=test_charm, release=[], name=None, format=formatted)
+
+        with pytest.raises(CraftError) as cm:
+            UploadCommand(config).run(args)
+        assert str(cm.value) == (
+            "Provided credentials are no longer valid for Charmhub. "
```

Comment:
what was the error before this fix?

Reply:
It should be a 401 unauthorized from here 
https://github.com/canonical/charmcraft/blob/9d0957ee4e7dda3327a93c79e0e0aff867c62498/charmcraft/commands/store/store.py#L152-L165

Reply:
Right, I mean, the purpose of this test is to check the early-failure right? How do we know that this failure is being caused by the whoami, and not by the rest of the upload process?

Reply:
Do we want to check traceback for this?

Reply:
Nah; can we mock the "heavy" call (looks like it might be `self._client.push_file()`?) and verify that it was _not_ called?

Reply:
OK, that works.